source/base/Text.ooc

```diff
@@ -0,0 +1,102 @@
+/*
+* Copyright (C) 2015 - Simon Mika <simon@mika.se>
+*
+* This sofware is free software; you can redistribute it and/or
+* modify it under the terms of the GNU Lesser General Public
+* License as published by the Free Software Foundation; either
+* version 2.1 of the License, or (at your option) any later version.
+*
+* This software is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	See the GNU
+* Lesser General Public License for more details.
+*
+* You should have received a copy of the GNU Lesser General Public License
+* along with this software. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+MemoryOwnership: enum {
+	Stack,
+	Literal,
+	Caller,
+	Callee
+}
+
+Text: cover {
+	_buffer: Char*
+	_length: Int
+	_memoryMode: MemoryOwnership
+	length ::= this _length
+	init: func@ ~fromStringLiteral (string: CString, =_length) {
+		this _buffer = string
+		this _memoryMode = MemoryOwnership Literal
+	}
+	init: func@ ~fromString (string: String) {
+		this _length = string length()
+		this _buffer = gc_malloc(this length)
+		memcpy(this _buffer, string _buffer data, this length)
+		this _memoryMode = MemoryOwnership Caller
+	}
+	empty?: func -> Bool {
+		this length == 0
+	}
+	startsWith?: func (other: This) -> Bool {
+		this _compareWithBuffer(other _buffer, other length, 0, 0) == 0
+	}
+	startsWith?: func ~string (other: String) -> Bool {
+		this _compareWithBuffer(other _buffer data, other length(), 0, 0) == 0
+	}
+	startsWith?: func ~char (c: Char) -> Bool {
+		if (this length > 0)
+			this _buffer[0] == c
+		false
+	}
+	endsWith?: func (other: This) -> Bool {
+		this _compareWithBuffer(other _buffer, other length, 0, this length - other length) == 0
+	}
+	endsWith?: func ~string (other: String) -> Bool {
+		this _compareWithBuffer(other _buffer data, other length(), 0, this length - other length()) == 0
+	}
+	endsWith?: func ~char (c: Char) -> Bool {
+		if (this length > 0)
+			this _buffer[this length - 1] == c
+		false
+	}
+	take: func@ {
+		this _changeOwnership(MemoryOwnership Caller)
+	}
+	give: func@ {
+		this _changeOwnership(MemoryOwnership Callee)
+	}
+	callerFree: func@ {
+		if (this _memoryMode == MemoryOwnership Caller) {
+			gc_free(this _buffer)
+			this _length = 0
+		}
+	}
+	calleeFree: func@ {
```

Comment:
Not that good names though, it should reflect that the *Calee* is done using this object and it can be safely released.